modules/ags/src/main/scala/lucuma/ags/Ags.scala

```diff
@@ -273,14 +274,34 @@ object Ags {
     // use constraints to calculate all guide speeds
     val bc    = constraintsFor(guideSpeeds)
 
-    candidates
-      .filter(c => c.gBrightness.exists(g => bc.exists(_.contains(Band.Gaia, g))))
-      .zip(positions.toList)
-      .map { (gsc, position) =>
-        val offset         = baseCoordinates.diff(gsc.tracking.baseCoordinates).offset
-        val scienceOffsets = scienceCoordinates.map(_.diff(gsc.tracking.baseCoordinates).offset)
-        runAnalysis(constraints, offset, scienceOffsets, position, params, gsc)(guideSpeeds, calcs)
-      }
+    // An optimal analysis result will be fast and deliver the requested IQ
+    def isOptimal(analysis: AgsAnalysis): Boolean = analysis match
+      case Usable(_, _, Some(GuideSpeed.Fast), AgsGuideQuality.DeliversRequestedIq, _, _) => true
+      case _                                                                              => false
+
+    @tailrec
+    def go(positions: List[AgsPosition], current: List[AgsAnalysis]): List[AgsAnalysis] = {
+      // We never pass an empty list
```

Comment:
Can we keep this information in the type, as a `NonEmptyList`?

We could do the recursion with `NonEmptyList.fromList(positions.tail).filterNot(_ => found).fold(<exit>)(<recurse>)`

Reply:
I was trying to avoid creating a NEL on each recursion step. Since it is a private method we can control the param is never empty

Reply:
I think using a NEL is not that expensive and it is safer even if it's private, in case we ever refactor. But I don't object to not using it.

Reply:
I'm not sure we can tail recurse when converting `positions.tail` to a `NEL`